src/proposal.jl

```diff
@@ -103,4 +110,38 @@ function q(
     t_cond
 )
     return q(proposal(t_cond), t, t_cond)
-end
\ No newline at end of file
+end
+
+"""
+    logratio_proposal_density(proposal, state, candidate)
+
+Compute the log-ratio of the proposal densities in the Metropolis-Hastings algorithm.
+
+The log-ratio of the proposal densities is defined as
+```math
+\\log \\frac{g(x | x')}{g(x' | x)},
+```
+where ``x`` is the current state, ``x'`` is the proposed candidate for the next state,
+and ``g(y' | y)`` is the conditional probability of proposing state ``y'`` given state
+``y`` (proposal density).
+"""
+function logratio_proposal_density(proposal::Proposal, state, candidate)
+    return q(proposal, state, candidate) - q(proposal, candidate, state)
+end
+
+# ratio is always 0 for symmetric random walk proposals
+logratio_proposal_density(::RandomWalkProposal{true}, state, candidate) = 0
```

Comment:
```suggestion
logratio_proposal_density(::RandomWalkProposal{true}, state, candidate) = 0.0
```
Not sure if `0` or `0.0` is better, one might create a type instability in sum reduction, the other might accidentally promote `Float32` to `Float64` (though we almost cannot do probability in `Float32` with the current design of `Distributions`)

Reply:
I was wondering the same but didn't check it. And indeed, it causes a type instability. However, returning a `Float64` here won't fix the underlying problem - the function will still be unstable. This is illustrated by the following example that shows that the type of the result can only be inferred if the array does not contain too many different proposals.

With `0`:
```julia
julia> @code_warntype AdvancedMH.logratio_proposal_density([RandomWalkProposal(Normal()), RandomWalkProposal{true}(Normal())], randn(2), randn(2))
Variables
  #self#::Core.Const(AdvancedMH.logratio_proposal_density)
  proposals::Vector{RandomWalkProposal{issymmetric, Normal{Float64}} where issymmetric}
  states::Vector{Float64}
  candidates::Vector{Float64}
  #73::AdvancedMH.var"#73#74"

Body::Union{Float64, Int64}
1 ─      (#73 = %new(AdvancedMH.:(var"#73#74")))
│   %2 = #73::Core.Const(AdvancedMH.var"#73#74"())
│   %3 = AdvancedMH.zip(proposals, states, candidates)::Base.Iterators.Zip{Tuple{Vector{RandomWalkProposal{issymmetric, Normal{Float64}} where issymmetric}, Vector{Float64}, Vector{Float64}}}
│   %4 = (:init,)::Core.Const((:init,))
│   %5 = Core.apply_type(Core.NamedTuple, %4)::Core.Const(NamedTuple{(:init,), T} where T<:Tuple)
│   %6 = Core.tuple(0)::Core.Const((0,))
│   %7 = (%5)(%6)::Core.Const((init = 0,))
│   %8 = Core.kwfunc(AdvancedMH.mapreduce)::Core.Const(Base.var"#mapreduce##kw"())
│   %9 = (%8)(%7, AdvancedMH.mapreduce, %2, AdvancedMH.:+, %3)::Union{Float64, Int64}
└──      return %9

julia> @code_warntype AdvancedMH.logratio_proposal_density([RandomWalkProposal(Normal()), RandomWalkProposal{true}(Normal())], randn(2), randn(2))
Variables
  #self#::Core.Const(AdvancedMH.logratio_proposal_density)
  proposals::Vector{RandomWalkProposal{issymmetric, Normal{Float64}} where issymmetric}
  states::Vector{Float64}
  candidates::Vector{Float64}
  #75::AdvancedMH.var"#75#76"

Body::Float64
1 ─      (#75 = %new(AdvancedMH.:(var"#75#76")))
│   %2 = #75::Core.Const(AdvancedMH.var"#75#76"())
│   %3 = AdvancedMH.zip(proposals, states, candidates)::Base.Iterators.Zip{Tuple{Vector{RandomWalkProposal{issymmetric, Normal{Float64}} where issymmetric}, Vector{Float64}, Vector{Float64}}}
│   %4 = (:init,)::Core.Const((:init,))
│   %5 = Core.apply_type(Core.NamedTuple, %4)::Core.Const(NamedTuple{(:init,), T} where T<:Tuple)
│   %6 = Core.tuple(0.0)::Core.Const((0.0,))
│   %7 = (%5)(%6)::Core.Const((init = 0.0,))
│   %8 = Core.kwfunc(AdvancedMH.mapreduce)::Core.Const(Base.var"#mapreduce##kw"())
│   %9 = (%8)(%7, AdvancedMH.mapreduce, %2, AdvancedMH.:+, %3)::Float64
└──      return %9

julia> @code_warntype AdvancedMH.logratio_proposal_density([RandomWalkProposal(Cauchy()), RandomWalkProposal{true}(Normal())], randn(2), randn(2))
Variables
  #self#::Core.Const(AdvancedMH.logratio_proposal_density)
  proposals::Vector{RandomWalkProposal}
  states::Vector{Float64}
  candidates::Vector{Float64}
  #75::AdvancedMH.var"#75#76"

Body::Any
1 ─      (#75 = %new(AdvancedMH.:(var"#75#76")))
│   %2 = #75::Core.Const(AdvancedMH.var"#75#76"())
│   %3 = AdvancedMH.zip(proposals, states, candidates)::Base.Iterators.Zip{Tuple{Vector{RandomWalkProposal}, Vector{Float64}, Vector{Float64}}}
│   %4 = (:init,)::Core.Const((:init,))
│   %5 = Core.apply_type(Core.NamedTuple, %4)::Core.Const(NamedTuple{(:init,), T} where T<:Tuple)
│   %6 = Core.tuple(0.0)::Core.Const((0.0,))
│   %7 = (%5)(%6)::Core.Const((init = 0.0,))
│   %8 = Core.kwfunc(AdvancedMH.mapreduce)::Core.Const(Base.var"#mapreduce##kw"())
│   %9 = (%8)(%7, AdvancedMH.mapreduce, %2, AdvancedMH.:+, %3)::Any
└──      return %9
```

With `0.0` type inference works in some cases but not reliably:
```julia
julia> @code_warntype AdvancedMH.logratio_proposal_density([RandomWalkProposal(Normal()), RandomWalkProposal{true}(Normal())], randn(2), randn(2))
Variables
  #self#::Core.Const(AdvancedMH.logratio_proposal_density)
  proposals::Vector{RandomWalkProposal{issymmetric, Normal{Float64}} where issymmetric}
  states::Vector{Float64}
  candidates::Vector{Float64}
  #75::AdvancedMH.var"#75#76"

Body::Float64
1 ─      (#75 = %new(AdvancedMH.:(var"#75#76")))
│   %2 = #75::Core.Const(AdvancedMH.var"#75#76"())
│   %3 = AdvancedMH.zip(proposals, states, candidates)::Base.Iterators.Zip{Tuple{Vector{RandomWalkProposal{issymmetric, Normal{Float64}} where issymmetric}, Vector{Float64}, Vector{Float64}}}
│   %4 = (:init,)::Core.Const((:init,))
│   %5 = Core.apply_type(Core.NamedTuple, %4)::Core.Const(NamedTuple{(:init,), T} where T<:Tuple)
│   %6 = Core.tuple(0.0)::Core.Const((0.0,))
│   %7 = (%5)(%6)::Core.Const((init = 0.0,))
│   %8 = Core.kwfunc(AdvancedMH.mapreduce)::Core.Const(Base.var"#mapreduce##kw"())
│   %9 = (%8)(%7, AdvancedMH.mapreduce, %2, AdvancedMH.:+, %3)::Float64
└──      return %9

julia> @code_warntype AdvancedMH.logratio_proposal_density([RandomWalkProposal(Cauchy()), RandomWalkProposal{true}(Normal())], randn(2), randn(2))
Variables
  #self#::Core.Const(AdvancedMH.logratio_proposal_density)
  proposals::Vector{RandomWalkProposal}
  states::Vector{Float64}
  candidates::Vector{Float64}
  #75::AdvancedMH.var"#75#76"

Body::Any
1 ─      (#75 = %new(AdvancedMH.:(var"#75#76")))
│   %2 = #75::Core.Const(AdvancedMH.var"#75#76"())
│   %3 = AdvancedMH.zip(proposals, states, candidates)::Base.Iterators.Zip{Tuple{Vector{RandomWalkProposal}, Vector{Float64}, Vector{Float64}}}
│   %4 = (:init,)::Core.Const((:init,))
│   %5 = Core.apply_type(Core.NamedTuple, %4)::Core.Const(NamedTuple{(:init,), T} where T<:Tuple)
│   %6 = Core.tuple(0.0)::Core.Const((0.0,))
│   %7 = (%5)(%6)::Core.Const((init = 0.0,))
│   %8 = Core.kwfunc(AdvancedMH.mapreduce)::Core.Const(Base.var"#mapreduce##kw"())
│   %9 = (%8)(%7, AdvancedMH.mapreduce, %2, AdvancedMH.:+, %3)::Any
└──      return %9
``` 

So to a large extent there are general limitation of arrays of proposals that one has to deal with. However, it even breaks for tuples in some cases (with `0`):

```julia

julia> @code_warntype AdvancedMH.logratio_proposal_density((RandomWalkProposal{true}(Normal()), RandomWalkProposal(Normal())), (randn(),randn()), (randn(),randn()))
Variables
  #self#::Core.Const(AdvancedMH.logratio_proposal_density)
  proposals::Tuple{SymmetricRandomWalkProposal{Normal{Float64}}, RandomWalkProposal{false, Normal{Float64}}}
  states::Tuple{Float64, Float64}
  candidates::Tuple{Float64, Float64}
  #113::AdvancedMH.var"#113#114"

Body::Union{Float64, Int64}
1 ─      (#113 = %new(AdvancedMH.:(var"#113#114")))
│   %2 = #113::Core.Const(AdvancedMH.var"#113#114"())
│   %3 = AdvancedMH.zip(proposals, states, candidates)::Base.Iterators.Zip{Tuple{Tuple{SymmetricRandomWalkProposal{Normal{Float64}}, RandomWalkProposal{false, Normal{Float64}}}, Tuple{Float64, Float64}, Tuple{Float64, Float64}}}
│   %4 = AdvancedMH.sum(%2, %3)::Union{Float64, Int64}
└──      return %4

julia> @code_warntype AdvancedMH.logratio_proposal_density((RandomWalkProposal(Cauchy()), RandomWalkProposal(Normal())), (randn(),randn()), (randn(),randn()))
Variables
  #self#::Core.Const(AdvancedMH.logratio_proposal_density)
  proposals::Tuple{RandomWalkProposal{false, Cauchy{Float64}}, RandomWalkProposal{false, Normal{Float64}}}
  states::Tuple{Float64, Float64}
  candidates::Tuple{Float64, Float64}
  #113::AdvancedMH.var"#113#114"

Body::Float64
1 ─      (#113 = %new(AdvancedMH.:(var"#113#114")))
│   %2 = #113::Core.Const(AdvancedMH.var"#113#114"())
│   %3 = AdvancedMH.zip(proposals, states, candidates)::Base.Iterators.Zip{Tuple{Tuple{RandomWalkProposal{false, Cauchy{Float64}}, RandomWalkProposal{false, Normal{Float64}}}, Tuple{Float64, Float64}, Tuple{Float64, Float64}}}
│   %4 = AdvancedMH.sum(%2, %3)::Float64
└──      return %4

julia> @code_warntype AdvancedMH.logratio_proposal_density((RandomWalkProposal(Cauchy()), RandomWalkProposal{true}(Normal())), (randn(),randn()), (randn(),randn()))
Variables
  #self#::Core.Const(AdvancedMH.logratio_proposal_density)
  proposals::Tuple{RandomWalkProposal{false, Cauchy{Float64}}, SymmetricRandomWalkProposal{Normal{Float64}}}
  states::Tuple{Float64, Float64}
  candidates::Tuple{Float64, Float64}
  #113::AdvancedMH.var"#113#114"

Body::Float64
1 ─      (#113 = %new(AdvancedMH.:(var"#113#114")))
│   %2 = #113::Core.Const(AdvancedMH.var"#113#114"())
│   %3 = AdvancedMH.zip(proposals, states, candidates)::Base.Iterators.Zip{Tuple{Tuple{RandomWalkProposal{false, Cauchy{Float64}}, SymmetricRandomWalkProposal{Normal{Float64}}}, Tuple{Float64, Float64}, Tuple{Float64, Float64}}}
│   %4 = AdvancedMH.sum(%2, %3)::Float64
└──      return %4

julia> @code_warntype AdvancedMH.logratio_proposal_density((StaticProposal(x -> Normal(x)), RandomWalkProposal{true}(Normal())), (randn(),randn()), (randn(),randn()))
Variables
  #self#::Core.Const(AdvancedMH.logratio_proposal_density)
  proposals::Tuple{StaticProposal{false, var"#15#16"}, SymmetricRandomWalkProposal{Normal{Float64}}}
  states::Tuple{Float64, Float64}
  candidates::Tuple{Float64, Float64}
  #113::AdvancedMH.var"#113#114"

Body::Union{}
1 ─      (#113 = %new(AdvancedMH.:(var"#113#114")))
│   %2 = #113::Core.Const(AdvancedMH.var"#113#114"())
│   %3 = AdvancedMH.zip(proposals, states, candidates)::Base.Iterators.Zip{Tuple{Tuple{StaticProposal{false, var"#15#16"}, SymmetricRandomWalkProposal{Normal{Float64}}}, Tuple{Float64, Float64}, Tuple{Float64, Float64}}}
│        AdvancedMH.sum(%2, %3)
└──      Core.Const(:(return %4))
```
Maybe one could unroll the iteration in some way or filter the proposals first, more similar to the current approach. I'll try different approaches.

Reply:
I solved a similar problem with https://github.com/tisztamo/FunctionWranglers.jl could be that that is overkill or too much code generation (it unrolls). Would be also ok to just force the computation in `Float64`

Reply:
IMO one shouldn't bother with the `Array`-case since the not concretely typed array of proposals leads to type inference problems in many other places as well (and is usually not what you want to use). As I assumed, `Tuple`s can be fixed with a recursive implementation, without enforcing `Float64`. I'll make sure that `NamedTuple`s are optimized as well.

Reply:
Perfect!

Reply:
I fixed the type inference problems.